Solutions/Menes.Hal/Menes/Hal/HalDocumentLinksCollectionWriter.cs

```diff
@@ -0,0 +1,316 @@
+﻿// <copyright file="HalDocumentLinksCollectionWriter.cs" company="Endjin Limited">
+// Copyright (c) Endjin Limited. All rights reserved.
+// </copyright>
+
+namespace Menes.Hal
+{
+    using System;
+    using System.Text.Json;
+
+    /// <summary>
+    /// Writer for the _links collection of a <see cref="HalDocumentResource"/>.
+    /// </summary>
+    public struct HalDocumentLinksCollectionWriter
```

Comment:
Everything's a `struct`!?

I find this one particularly odd because the general rule with structs is to use them for value-like things; using them for entities with their own significant identity is problematic, because it's so easy to end up working accidentally with a copy. (C# frequently takes the attitude that it's generally fine to work with a copy of the value, except in certain quite narrow scenarios.)

So modelling a thing with its own state as a `struct` seems like a recipe for disaster. Why is this a `struct`?

Reply:
The goal (ultimately) is to end up with this lot all allocated on the stack in your handler path, and we avoid GC allocations.